src/openrct2/world/Footpath.cpp

```diff
@@ -1272,148 +1272,210 @@ static bool get_next_direction(int32_t edges, int32_t* direction)
+                continue;
+
+            // Unown the footpath if needed
+            if (flags & FOOTPATH_CONNECTED_MAP_EDGE_UNOWN)
+                footpath_fix_ownership(targetPos);
+
+            edges = tileElement->AsPath()->GetEdges();
+            direction = direction_reverse(direction);
+            if (!(flags & FOOTPATH_CONNECTED_MAP_EDGE_IGNORE_NO_ENTRY))
             {
-                for (int32_t i = 1; i < 6; i++)
+                // Determine how many elements to check
+                size_t limit = 0;
+                if (tileElement[1].GetType() == TILE_ELEMENT_TYPE_BANNER)
+                    limit = 4;
+                if (tileElement[2].GetType() == TILE_ELEMENT_TYPE_BANNER && tileElement[1].GetType() != TILE_ELEMENT_TYPE_PATH)
+                    limit = 6;
+
+                for (size_t ii(1); ii < limit; ++ii)
                 {
-                    if ((&tileElement[i - 1])->IsLastForTile())
+                    if ((&tileElement[ii - 1])->IsLastForTile())
                         break;
-                    if (tileElement[i].GetType() != TILE_ELEMENT_TYPE_BANNER)
+                    if (tileElement[ii].GetType() != TILE_ELEMENT_TYPE_BANNER)
                         break;
-                    edges &= tileElement[i].AsBanner()->GetAllowedEdges();
+                    edges &= tileElement[ii].AsBanner()->GetAllowedEdges();
                 }
             }
-        }
-        goto searchFromFootpath;
-    } while (!(tileElement++)->IsLastForTile());
-    return level == 1 ? FOOTPATH_SEARCH_NOT_FOUND : FOOTPATH_SEARCH_INCOMPLETE;
-
-searchFromFootpath:
-    // Exclude direction we came from
-    targetPos.z = tileElement->GetBaseZ();
-    edges &= ~(1 << direction);
 
-    // Find next direction to go
-    int32_t newDirection{};
-    if (!get_next_direction(edges, &newDirection))
-    {
-        return FOOTPATH_SEARCH_INCOMPLETE;
-    }
-    direction = newDirection;
+            // Exclude the direction we came from
+            targetPos.z = tileElement->GetBaseZ();
+            edges &= ~(1 << direction);
 
-    edges &= ~(1 << direction);
-    if (edges == 0)
-    {
-        // Only possible direction to go
-        if (tileElement->AsPath()->IsSloped() && tileElement->AsPath()->GetSlopeDirection() == direction)
-        {
-            targetPos.z += PATH_HEIGHT_STEP;
-        }
-        return footpath_is_connected_to_map_edge_recurse(
-            targetPos, direction, flags, level, distanceFromJunction + 1, junctionTolerance);
-    }
-    else
-    {
-        // We have reached a junction
-        if (distanceFromJunction != 0)
-        {
-            junctionTolerance--;
-        }
-        junctionTolerance--;
-        if (junctionTolerance < 0)
-        {
-            return FOOTPATH_SEARCH_TOO_COMPLEX;
-        }
+            // Find next direction to go
+            if (!get_next_direction(edges, &direction))
+                continue;
```

Comment:
This continue looks incorrect. You are going to repeat the do while but shouldn't this exit the do while and do another iteration of the while.

Reply:
I believe you are correct; each separate pass of the do-while loop should correspond to a single recursive call in the original code, so where that return was should most certainly be a `break` rather than a `continue`.